Latest macOS system header causes compilation failures on GCC

**Description**
After recent developer tools update on macOS Sonoma (i.e. "Command Line Tools for Xcode 15.3"), some Catch2 headers fail to compile on GCC.

**Test case**
```
// v2's `single_include/catch2/catch.hpp` includes `TargetConditionals.h`
#include <catch2/catch.hpp>
```
and
```
// v3's `catch2/internal/catch_platform.hpp`, which is included by `catch2/catch_all.hpp`, includes `TargetConditionals.h`
#include <catch2/catch_all.hpp>
```
fail to compile on GCC 13 or earlier, after updating to "Command Line Tools for Xcode 15.3". 

**Background**
"Command Line Tools for Xcode 15.3" introduces a problematic code at Line 140 of `usr/include/TargetConditionals.h` (i.e. `MacOSX14.4.sdk/usr/include/TargetConditionals.h`):
```
#if !defined(__has_extension) || # !__has_extension(define_target_os_macros)
```
This is problematic, since compilers without `__has_extension` support cause compilation failure with it. Notably, GCC does not support `__has_extension` until the next release (GCC 14).

(I think this should be fixed upstream by Apple, but ...) it may make some sense to add guard for `__has_extension`:
```
#ifdef __APPLE__
#  ifndef __has_extension
#  define __has_extension(x) 0
#  endif
#  include <TargetConditionals.h>
```

Yes, that's hard errors. For example, on mac with the latest developer tools, this
```
echo '#include <catch2/catch_all.hpp>' | g++-13 -xc++ -
```
produces the following error messages
```
In file included from catch2/internal/catch_platform.hpp:14,
                 from catch2/internal/catch_compiler_capabilities.hpp:27,
                 from catch2/benchmark/catch_benchmark.hpp:14,
                 from catch2/benchmark/catch_benchmark_all.hpp:24,
                 from catch2/catch_all.hpp:25,
                 from <stdin>:1:
/usr/include/TargetConditionals.h:140:50: error: missing binary operator before token "("
  140 | #if !defined(__has_extension) || !__has_extension(define_target_os_macros)
```
The code `!__has_extension(define_target_os_macros)` causes compilation failure, because `__has_extension` is not supported/defined on GCC (until the next release of GCC).

--- a/src/catch2/internal/catch_platform.hpp
+++ b/src/catch2/internal/catch_platform.hpp
@@ -11,6 +11,9 @@
 // See e.g.:
 // https://opensource.apple.com/source/CarbonHeaders/CarbonHeaders-18.1/TargetConditionals.h.auto.html
 #ifdef __APPLE__
+#  ifndef __has_extension
+#    define __has_extension(x) 0
+#  endif
 #  include <TargetConditionals.h>
 #  if (defined(TARGET_OS_OSX) && TARGET_OS_OSX == 1) || \
       (defined(TARGET_OS_MAC) && TARGET_OS_MAC == 1)

